### PR TITLE
[recipe] feat: CollabLLM integration for multiturn training

### PR DESCRIPTION
**Pull request by Wuyxin**
### What does this PR do?

This PR add [CollabLLM](https://aka.ms/CollabLLM) as a training recipe. The added components include 
- A customized `CollabLLMRewardManager` inheriting from `AbstractRewardManager` to compute multiturn-aware rewards.
- A customized `CollabLLMAgentLoop` inheriting from `AgentLoop` to sample future conversations with simulated users, which imports `CollabLLMInteraction` from `verl/interactions/collabllm_interation.py`.

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

The training rewards when running `train_rl_collabllm.sh` is increasing in a relatively stable manner (on 8xH200):
9baeb0700e3fa6a56596e14a54bc1049

Validation reward:
39364fd10523b0fde13d48645809f5e3

#### Samples of model generation
After training, when user asks generic questions with missing information, the model learns to ask for clarification
c8e0ab31948a48ca396c7eccddd13673
and give suggestions:
7adb7d33eb9120d337c2a249c6a2dd22

(In contrast, with the same prompt, **GPT-5** doesn't ask for any clarification:)
be8d8577584c0b2356cb352d6f294205

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# No change on the existing APIs
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

Changes:
- Main files under `recipe/collabllm`
- Registered `CollabLLMRewardManager` in `workers/reward_manager/collabllm.py`
- Added `CollabLLMInteraction` in `verl/interactions/collabllm_interation.py`

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs). Added to `verl/docs/algo/collabllm.md`.
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: The scripts `train_rl_collabllm.sh` and `train_sft_collabllm.sh` are tested multiple times.
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/workers/reward_manager/collabllm.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            score_dicts = []
+            scores = torch.full((batch_size,), 0.0, dtype=torch.float32, device=prompt_ids.device)
+            mean_weighted_scores_by_metrics = {metric: 0.0 for metric in self.metrics}
+
+        print("Scores:", scores, mean_weighted_scores_by_metrics)
+
+        reward_tensor = torch.zeros_like(data.batch["responses"], dtype=torch.float32)
+
+        for i in range(len(data)):
+            reward_tensor[i, valid_response_length[0].item() - 1] = scores[i]
```
[the remark this answers is not in the document]
Comment (Wuyxin) [reply verl/workers/reward_manager/collabllm.py]: already fixed

File: verl/workers/rollout/sglang_rollout/sglang_rollout.py [hunk 2]
```diff
@@ -1104,55 +1114,98 @@ def _req_level_generate_sequences(self, prompts: DataProto, **kwargs) -> DataPro
...
-                        logger.info(f"Request {req.request_id} was cancelled, creating padding")
-                        aborted_requests.append(req.request_id)
-                        return self._create_padding_request(req)
-
-                async def run_with_cancellation():
-                    nonlocal all_tasks
-                    nonlocal completed_count
-                    all_tasks = [
-                        asyncio.create_task(rollout_a_request_with_cancellation_handler(req)) for req in req_list
+                if self.config.multi_turn.collabllm_rollouts:
```
Comment (wuxibin89) [inline verl/workers/rollout/sglang_rollout/sglang_rollout.py]: As we're moving to agent loop, these changes should not be needed.

**Comment by wuxibin89** [comment]
@Wuyxin Hi, thanks for your contribution, please cleanup irrelevant code, e.g `recipe/char_count`, `recipe/prime`, `sglang_rollout.py`.

**Comment by Wuyxin** [comment]
Thanks for the quick review.

I have removed the accidental chances and the `if self.config.multi_turn.collabllm_rollouts:` branch in `sglang_rollouts.py`, as well as the flag `collabllm_rollouts`.

File: verl/workers/reward_manager/collabllm.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+from verl import DataProto
+from verl.utils.reward_score import default_compute_score
+from verl.workers.reward_manager import register
+from verl.workers.reward_manager.abstract import AbstractRewardManager
+
+TERMINATION_SIGNAL = "[[TERMINATE CHAT]]"
+
+
+@register("collabllm")
+class CollabLLMRewardManager(AbstractRewardManager):
```
Comment (wuxibin89) [inline verl/workers/reward_manager/collabllm.py:31]: P2: We better move this to recipe and add a option in config to load custom reward manager. Leave it for now, we will refactor it later.

**Comment by wuxibin89** [comment]
@Wuyxin @chenhaiq Can we move these two files to `recipe/colabllm`?
- verl/interactions/collabllm_interation.py
- verl/workers/reward_manager/collabllm.py
